### DISCUSSION TITLE
Buildroot + Kconfig + RAUC

### DISCUSSION THREAD

**Discussion by Sourabh1107** [rauc/rauc#798] [Q&A]
Hi,

I am integrating RAUC with Buildroot for one of the IMX8MM boards. I understand that we need to have support for loop block devices and the SquashFS file system. And to integrate with RAUC we have enabled BR2_PACKAGE_RAUC=y, BR2_PACKAGE_RAUC_NETWORK=y, BR2_PACKAGE_RAUC_JSON=y in the configuration. But my concern is how to enable the flags in the Kconfig as mentioned in the RAUC documentation. Can we directly go ahead and add these in our custom configuration file (along with  BR2_PACKAGE_RAUC=y)? How will the dependency be handled if there are any? 

Please let me know if someone has successfully integrated RAUC with Buildroot.

**Comment by hongkongkiwi** [comment]
**TLDR:** use `make menuconfig` in buildroot, everything (except kernel config) is automatically setup for you, no need to add or edit anything by hand.

**Explanation**
First make sure you are setting up your config in buildroot with `make menuconfig` command, instead of editing by hand. The menu system is really the recommended way to set things up in buildroot. Note that in buildroot they are not actually Kconfig files, but called Config.in in most packages.

In the example of RAUC inside your buildroot directory there is a package definition `package/rauc/Config.in`. This file contains the kconfig style file which specifies the menu option.

For example, you can see in this file:
```
config BR2_PACKAGE_RAUC_NETWORK
  bool "network support"
  select BR2_PACKAGE_LIBCURL
  help
    This option enables support for updating firmware over
    the network using libcurl.
```

This means that when you select "network support" option in the menu, it will select the package libcurl (BR2_PACKAGE_LIBCURL) as a dependency. This should happen automatically when you select the option.

If you look at the top of the file it looks like this:
```
config BR2_PACKAGE_RAUC
  bool "rauc"
  depends on BR2_TOOLCHAIN_HAS_THREADS # glib2
  depends on BR2_USE_MMU # glib2
  depends on BR2_USE_WCHAR # glib2
  select BR2_PACKAGE_LIBGLIB2
  select BR2_PACKAGE_OPENSSL
  select BR2_PACKAGE_DBUS
  select BR2_PACKAGE_SQUASHFS # run-time dependency
  select BR2_PACKAGE_UBOOT_TOOLS if BR2_TARGET_UBOOT # run-time dependency
  select BR2_PACKAGE_UBOOT_TOOLS_FWPRINTENV if BR2_TARGET_UBOOT
  help
    RAUC is the Robust Auto-Update Controller developed by
    Pengutronix. It supports updating embedded Linux
    systems over the network or from disks and provides a
    d-bus interface.

    http://rauc.io/
```

So now you can see that when you select the RAUC package it automatically selects a number of packages, e.g. openssl, uboot-tools etc.

Hopefully it now makes sense how the sub packages are selected.

The only thing that isn't automatically selected is the kernel options you need to select, in my case the kernel is built separately so buildroot can't automatically select those.

**Comment by Sourabh1107** [comment]
Hi @hongkongkiwi , Thanks for the quick response. Yes I understood that the dependencies are automatically selected. That's great. And also we need to setup kernel config separately. In the RAUC documentation [https://rauc.readthedocs.io/en/latest/integration.html#kernel-configuration] its mentioned that "In kernel Kconfig you have to enable the following options:
CONFIG_MD=y
CONFIG_BLK_DEV_DM=y
CONFIG_BLK_DEV_LOOP=y
CONFIG_DM_VERITY=y
CONFIG_SQUASHFS=y
CONFIG_CRYPTO_SHA256=y
"
And I believe these options are enabled by using `make linux-menuconfig`. Any idea which options correspond to these settings in kernel config menu? I am trying to understand these options in buildroot.

**Reply by jluebbe** [reply]
You can use `/` to search for options in menuconfig (without the `CONFIG_` prefix, i.e. `/DM_VE`).

**Reply by ejoerns** [reply]
@Sourabh1107  The above-mentioned options actually are the exact names in the kernel kconfig, buildroot will surely also only call the kernel menuconfig directly and not do any mapping. Have you been able to find it? If so, please leave a comment and/or mark this as answered, thx :)

**Comment by jluebbe** [comment]
I'll move this to the discussion section, as this is not an issue in RAUC itself.
